hera_stats/jackknives.py

```diff
@@ -0,0 +1,353 @@
+import copy
+import os
+import utils
+import astropy.coordinates as cp
+
+def _bootstrap_uvp_once(uvp, pol="xx", n_boots=100):
+    
+    """
+    Uses the bootstrap method to estimate the errors for a PSpecData object. 
+
+    Parameters:
+    ----------
+    uvp: hera_pspec.UVPSpec
+        Object outputted by pspec, contains power spectrum information.
+
+    pol: str, optional
+        The polarization used in pspec calculations. Default: "xx".
+
+    n_boots: int, optional
+        How many bootstraps of the data to to to estimate error bars. Default: 100.
+
+    Returns:
+    -------
+    uvp_avg: hera_pspec.UVPSpec
+        A UVPSpec containing the averaged power spectrum and bootstrapped errors.
+    """
+    # Calculate unique baseline pairs
+    blpairs = uvp.get_blpairs()
+    uvp_avg = uvp.average_spectra([blpairs], time_avg = True, inplace=False)
+
+    for spw in range(uvp.Nspws):
+        allspecs = uvp.data_array[spw]
+        nsamples = uvp.nsample_array[spw]
+        integs = uvp.integration_array[spw]
+
+        boot_l = []
+        for i in range(n_boots):
+            # Choose spectra indices at random
+            inds = np.random.choice(range(len(allspecs)), len(allspecs), replace=True)
+            
+            # Select spectra, nsamples, and integrations in order.
+            boot = allspecs[inds]
+            nsamps = nsamples[inds]
+            ints = integs[inds]
+
+            boot_uvp = copy.deepcopy(uvp)
+
+            # Replace old arrays with new ones.
+            boot_uvp.data_array[spw] = boot
+            boot_uvp.nsample_array[spw] = nsamps
+            boot_uvp.integration_array[spw] = ints
+
+            # Average spectra
+            boot_uvp.average_spectra([blpairs], time_avg=True)
+
+            # Get data and append
+            boot_spec = boot_uvp.get_data((spw, uvp.get_blpairs()[0], pol))[0]
+            boot_l.append(boot_spec)
+
+        # Calculate error bar and save to stats_array
+        boot_l = np.vstack(boot_l)
+        errs = np.std(boot_l.real, axis=0) + 1j * np.std(boot_l.imag, axis=0)
+        uvp_avg.set_stats("bootstrap_errs", (spw, uvp.get_blpairs()[0], pol), errs[None])
+
+    return uvp_avg
+
+def bootstrap_jackknife(uvpl, pol="xx", n_boots=50):
+    """
+    Bootstraps a list of jackknife pairs.
+
+    Parameters
+    ----------
+    uvpl: list of UVPSpec pairs
+        The jackknived data, with UVPSpec.labels as the group and UVPSpec.jkftype as 
+        the jackknife type, shape=(n, 2).
+
+    pol: string, optional
+        Polarization to use for bootstrapping. Default="xx".
+
+    n_boots: int, optional
+        Number of times to bootstrap to create a sample. Default: 50.
+    """
+    data = []
+    for uv in uvpl:
+        bspair = []
+        for u in uv:
+            uc = copy.deepcopy(u)
+            bspair.append(_bootstrap_uvp_once(uc, pol=pol, n_boots=n_boots))
+        data.append(bspair)
+    return data
+
+def save_jackknife(uvpl, savename=None):
+    """
```

Comment:
Can you make it so that this can save into a user-specified group in an existing PSpecContainer? The user can just pass a handle to a HDF5 group.